contracts/misc/AutoYieldApe.sol

```diff
@@ -0,0 +1,306 @@
+        emit RescueERC20(token, to, amount);
+    }
+
+    function _withdraw(uint256 amount) internal {
+        require(amount > 0, "zero amount");
+
+        _updateYieldIndex(msg.sender);
+        _burn(msg.sender, amount);
+
+        _apeStaking.withdrawSelfApeCoin(amount);
+        IERC20(_apeCoin).safeTransfer(msg.sender, amount);
+
+        emit Redeem(msg.sender, amount);
+    }
+
+    function _claim() internal {
+        uint256 pendingYield = yieldAmount(msg.sender);
+        if (pendingYield > 0) {
+            _userPendingYield[msg.sender] = 0;
+            IERC20(_yieldToken).safeTransfer(msg.sender, pendingYield);
+        }
+        emit YieldClaimed(msg.sender, pendingYield);
+    }
+
+    function _harvest(uint160 sqrtPriceLimitX96) internal {
+        uint256 rewardAmount = _apeStaking.pendingRewards(
+            APE_COIN_POOL_ID,
+            address(this),
+            0
+        );
+        if (rewardAmount > 0) {
+            _apeStaking.claimSelfApeCoin();
+            uint256 yieldUnderlyingAmount = _sellApeCoinForYieldToken(
+                rewardAmount,
+                sqrtPriceLimitX96
+            );
+            IPoolCore(_lendingPool).supply(
+                _yieldUnderlying,
+                yieldUnderlyingAmount,
+                address(this),
+                0
+            );
+            uint256 liquidityIndex = _lendingPool.getReserveNormalizedIncome(
+                _yieldUnderlying
+            );
+            uint256 _yieldAmount = yieldUnderlyingAmount.rayDiv(liquidityIndex);
+            uint256 accuIndex = (_yieldAmount * RAY) / totalSupply();
+            _currentYieldIndex += accuIndex;
+        }
+    }
+
+    function _updateYieldIndex(address account) internal {
+        uint256 currentYieldIndex = _currentYieldIndex;
+        uint256 indexDiff = currentYieldIndex - _userYieldIndex[account];
+        if (indexDiff > 0) {
+            uint256 rewardDiff = (balanceOf(account) * indexDiff) / RAY;
+            if (rewardDiff > 0) {
+                _userPendingYield[account] += rewardDiff;
+            }
+            _userYieldIndex[account] = currentYieldIndex;
+        }
+    }
+
+    function _sellApeCoinForYieldToken(
+        uint256 apeCoinAmount,
+        uint160 sqrtPriceLimitX96
+    ) internal returns (uint256) {
+        return
+            _swapRouter.exactInputSingle(
+                ISwapRouter.ExactInputSingleParams({
+                    tokenIn: _apeCoin,
+                    tokenOut: _yieldUnderlying,
+                    fee: 3000,
+                    recipient: address(this),
+                    deadline: block.timestamp,
+                    amountIn: apeCoinAmount,
+                    amountOutMinimum: 0,
+                    sqrtPriceLimitX96: sqrtPriceLimitX96
+                })
+            );
+    }
+
+    function _transfer(
+        address sender,
+        address recipient,
+        uint256 amount
+    ) internal override {
+        _updateYieldIndex(sender);
```

Comment:
normal erc20 assumes that sender can be equal to recipient, shall we check first if it's equal